librz/main/rz-diff.c

```diff
@@ -1465,100 +1480,97 @@ typedef enum diff_hex_len_t {
 	for (i = 0; i < hexlen && i < read_a; ++i) {
-		if (i < read_b) {
-			same = buffer_a[pos + i] == buffer_b[pos + i];
-			lp += printline("%s%02x%s ", same ? match : unmatch, buffer_a[pos + i], reset);
+		if (pos + i >= size_a || pos + i < skip_a) {
+			// if the byte is outside the range [0 - fileA size) then do not write any hex
+			memset(line + lp, ' ', 3);
+			lp += 3;
+		} else if (i < read_b && pos + i >= skip_b) {
+			// if the byte is inside the range [0 - fileA size) check bytes_b for match/mismatch
+			const char *color = bytes_a[pos + i] == bytes_b[pos + i] ? match : unmatch;
+			lp += printline("%s%02x%s ", color, bytes_a[pos + i], reset);
 		} else {
-			lp += printline("%s%02x%s ", unmatch, buffer_a[pos + i], reset);
+			// if the byte is inside the range [0 - fileA size) but address_b
+			// is outside [0 - fileB size) then is a mismatch
+			lp += printline("%s%02x%s ", unmatch, bytes_a[pos + i], reset);
 		}
 	}
 	if (i < hexlen) {
+		// fill any missing space to have fileA bytes aligned
 		memset(line + lp, ' ', (hexlen - i) * 3);
 		lp += (hexlen - i) * 3;
 	}
 
+	// print now printable chars of the printed hex values
 	lp += printline(" | ");
 	for (i = 0; i < hexlen && i < read_a; ++i) {
-		if (i < read_b) {
-			same = buffer_a[pos + i] == buffer_b[pos + i];
-			lp += printline("%s%c%s", same ? match : unmatch, P(buffer_a[pos + i]), reset);
+		if (pos + i >= size_a || pos + i < skip_a) {
+			// if the byte is outside the range [0 - fileA size) then do not write any hex
+			line[lp] = ' ';
+			lp++;
+		} else if (i < read_b && pos + i >= skip_b) {
+			// if the byte is inside the range [0 - fileA size) check bytes_b for match/mismatch
+			const char *color = bytes_a[pos + i] == bytes_b[pos + i] ? match : unmatch;
+			lp += printline("%s%c%s", color, P(bytes_a[pos + i]), reset);
 		} else {
-			lp += printline("%s%c%s", unmatch, P(buffer_a[pos + i]), reset);
+			// if the byte is inside the range [0 - fileA size) but address_b
+			// is outside [0 - fileB size) then is a mismatch
+			lp += printline("%s%c%s", unmatch, P(bytes_a[pos + i]), reset);
 		}
 	}
 	if (i < hexlen) {
+		// fill any missing space to have fileA bytes aligned
 		memset(line + lp, ' ', (hexlen - i));
 		lp += (hexlen - i);
 	}
+	return lp;
+#undef printline
+#undef P
+}
 
-	lp += printline(" | %s0x%016" PFMT64x "%s | ", number, offset_b + pos, reset);
-	for (i = 0; i < hexlen && i < read_b; ++i) {
-		if (i < read_a) {
-			same = buffer_a[pos + i] == buffer_b[pos + i];
-			lp += printline("%s%02x%s ", same ? match : unmatch, buffer_b[pos + i], reset);
-		} else {
-			lp += printline("%s%02x%s ", unmatch, buffer_b[pos + i], reset);
-		}
-	}
-	if (i < hexlen) {
-		memset(line + lp, ' ', (hexlen - i) * 3);
-		lp += (hexlen - i) * 3;
-	}
+static inline void diff_hexdump_line(DiffHexView *hview, DiffHexLen hlen, ut64 pos, ssize_t read_a, ssize_t read_b, ssize_t skip_a, ssize_t skip_b) {
```

Comment:
Is it possible to avoid so many arguments? Maybe combine them into some kind of structure?

Reply:
it's an inline function, there is no real difference on using a struct or not in this specific case.